pyccl/cells.py

```diff
@@ -69,17 +110,53 @@ def angular_cl(cosmo, tracer1, tracer2, ell, *,
     if not (np.diff(ell_use) > 0).all():
         raise ValueError("ell values must be monotonically increasing")
 
+    if auto_limber or (type(l_limber) is not str and ell_use[0] < l_limber):
+        if non_limber_integration_method == "FKEM":
+            l_limber, cl_non_limber, status = nonlimber_FKEM(
+                cosmo,
+                tracer1,
+                tracer2,
+                p_of_k_a,
+                p_of_k_a_lin,
+                ell_use,
+                l_limber,
+                limber_max_error,
+            )
+        # else:  # it has to be matter, since we checked the input
+        #     l_limber, cl_non_limber, status = implement_MATTER(
+        #         cosmo, clt1, clt2, psp, ell_use, l_limber, limber_max_error
+        #     )
+        check(status, cosmo=cosmo_in)
+    else:
+        cl_non_limber = np.array([])
+    cosmo = cosmo.cosmo
+    ell_use_limber = ell_use[ell_use > l_limber]
     # Return Cl values, according to whether ell is an array or not
-    cl, status = lib.angular_cl_vec(
-        cosmo, clt1, clt2, psp, l_limber,
-        ell_use, integ_types[limber_integration_method],
-        ell_use.size, status)
+    if len(ell_use_limber) > 0:
+        cl_limber, status = lib.angular_cl_vec_limber(
+            cosmo,
+            clt1,
+            clt2,
+            psp,
+            ell_use_limber,
+            integ_types[limber_integration_method],
+            ell_use_limber.size,
+            status,
+        )
+    else:
+        cl_limber = np.array([])
+
+    # put pieces together
+    cl = np.concatenate((cl_non_limber, cl_limber))
     if np.ndim(ell) == 0:
         cl = cl[0]
 
     # Free up tracer collections
     lib.cl_tracer_collection_t_free(clt1)
     lib.cl_tracer_collection_t_free(clt2)
 
+    if return_meta:
+        meta = {"l_limber": l_limber}  # add other things as needed
```

Comment:
the user may enter 'auto', meaning the non-limber calculation will end after the limber/non-limber calculations are sufficiently close to one another (also set by the user). This will return the l-value indicating where limber calculations begin

Reply:
right, great.